File: skills/retrospect/scripts/parse_session.py
#!/usr/bin/env python3
"""Parse AI coding assistant session JSONL files into readable Markdown transcripts.

Supports: Claude Code, Codex (OpenAI), OpenCode, and generic JSONL chat formats.

Usage:
    # Parse all sessions for a project (auto-discovers JSONL files):
    python3 parse_session.py --project-dir /path/to/project

    # Parse a single JSONL file:
    python3 parse_session.py /path/to/session.jsonl

Outputs Markdown to stdout. Redirect to a file as needed.
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path

# Older sessions beyond this count are summarized (user messages only)
# to keep the transcript manageable. Most recent sessions get full detail.
FULL_DETAIL_SESSIONS = 5
MAX_SUMMARY_TURNS = 10


def extract_text_from_content(content):
    """Extract readable text from message content (string or content blocks)."""
    if isinstance(content, str):
        return content.strip()
    if isinstance(content, list):
        parts = []
        for block in content:
            if not isinstance(block, dict):
                continue
            btype = block.get("type", "")
            if btype == "text":
                parts.append(block["text"].strip())
            elif btype == "tool_use":
                name = block.get("name", "unknown")
                inp = block.get("input", {})
                if name in ("Read", "Glob", "Grep", "read_file", "grep", "glob"):
                    target = inp.get("file_path") or inp.get("pattern") or inp.get("path", "")
                    parts.append(f"[Tool: {name}({target})]")
                elif name in ("Bash", "shell", "run_command"):
                    cmd = inp.get("command", "")
                    if len(cmd) > 200:
                        cmd = cmd[:200] + "..."
                    parts.append(f"[Tool: {name}] {cmd}")
                elif name in ("Write", "Edit", "write_file", "edit_file"):
                    fp = inp.get("file_path", "")
                    parts.append(f"[Tool: {name}({fp})]")
                elif name in ("Agent", "spawn_agent"):
                    desc = inp.get("description", inp.get("prompt", ""))[:100]
                    parts.append(f"[Tool: Agent({desc})]")
                else:
                    parts.append(f"[Tool: {name}]")
            elif btype == "tool_result":
                result_content = block.get("content", "")
                if isinstance(result_content, str) and result_content.strip():
                    text = result_content.strip()
                    if len(text) > 300:
                        text = text[:300] + "... [truncated]"
                    parts.append(f"[Tool Result] {text}")
                elif isinstance(result_content, list):
                    for sub in result_content:
                        if isinstance(sub, dict) and sub.get("type") == "text":
                            text = sub["text"].strip()
                            if len(text) > 300:
                                text = text[:300] + "... [truncated]"
                            parts.append(f"[Tool Result] {text}")
            elif btype == "function_call":
                fname = block.get("name", "unknown")
                parts.append(f"[Function: {fname}]")
            elif btype == "thinking":
                pass
        return "\n".join(parts)
    return ""


def detect_format(obj):
    """Detect JSONL format from a sample object."""
    if "type" in obj and obj["type"] in ("user", "assistant", "system", "file-history-snapshot"):
        return "claude-code"
    if "role" in obj and "content" in obj:
        return "openai"
    return "unknown"


def parse_claude_code(lines):
    """Parse Claude Code JSONL format."""
    exchanges = []
    for line in lines:
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue

        msg_type = obj.get("type", "")

        if msg_type in ("file-history-snapshot", "attachment"):
            continue
        if obj.get("isMeta"):
            continue

        if msg_type == "user":
            content = obj.get("message", {}).get("content", "")
            text = extract_text_from_content(content)
            if text and "<command-name>" in text:
                continue
            if text and not text.startswith("[Tool Result]"):
                exchanges.append(("user", text))
            elif text:
                exchanges.append(("tool_result", text))

        elif msg_type == "assistant":
            content = obj.get("message", {}).get("content", [])
            text = extract_text_from_content(content)
            if text:
                exchanges.append(("assistant", text))

        elif msg_type == "system":
            content = obj.get("content", "")
            if content and len(content.strip()) > 10:
                subtype = obj.get("subtype", "")
                if subtype != "local_command":
                    exchanges.append(("system", content.strip()[:500]))

    return exchanges


def parse_openai_format(lines):
    """Parse OpenAI/Codex/OpenCode JSONL format (role/content pairs)."""
    exchanges = []
    for line in lines:
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue

        role = obj.get("role", "")
        content = obj.get("content", "")
        text = extract_text_from_content(content)

        if not text:
            continue

        if role in ("user", "human"):
            exchanges.append(("user", text))
        elif role in ("assistant", "model"):
            exchanges.append(("assistant", text))
        elif role == "system":
            if len(text) > 10:
                exchanges.append(("system", text[:500]))
        elif role == "tool":
            if len(text) > 10:
                exchanges.append(("tool_result", text[:300]))

    return exchanges


def parse_single_file(jsonl_path):
    """Parse a single JSONL file. Auto-detects format."""
    lines = Path(jsonl_path).read_text().strip().split("\n")

    fmt = "unknown"
    for line in lines:
        try:
            obj = json.loads(line)
            fmt = detect_format(obj)
            if fmt != "unknown":
                break
        except json.JSONDecodeError:
            continue

    if fmt == "claude-code":
        return parse_claude_code(lines)
    elif fmt == "openai":
        return parse_openai_format(lines)
    else:
        cc = parse_claude_code(lines)
        oai = parse_openai_format(lines)
        return cc if len(cc) >= len(oai) else oai


def summarize_exchanges(exchanges):
    """Summarize older sessions: keep only user messages, limit count."""
    summary = []
    count = 0
    for role, text in exchanges:
        if role == "user":
            count += 1
            if count > MAX_SUMMARY_TURNS:
                summary.append(("user", f"... [{count}+ more user turns omitted] ..."))
                break
            # Truncate long user messages in summary
            if len(text) > 300:
                text = text[:300] + "... [truncated]"
            summary.append(("user", text))
    return summary


def build_markdown_multi(session_list):
    """Build Markdown from multiple sessions.

    session_list: list of (session_file, mtime, exchanges) sorted oldest-first.
    Older sessions beyond FULL_DETAIL_SESSIONS are summarized.
    """
    output = ["# Project Session Transcripts\n"]
    total = len(session_list)
    full_start = max(0, total - FULL_DETAIL_SESSIONS)

    for i, (session_file, mtime, exchanges) in enumerate(session_list):
        if not exchanges:
            continue

        session_num = i + 1
        date_str = datetime.fromtimestamp(mtime).strftime("%Y-%m-%d %H:%M")
        session_id = Path(session_file).stem[:8]

        output.append(f"---\n")
        output.append(f"## Session {session_num} ({date_str}, id: {session_id}...)\n")

        if i < full_start:
            # Summarize older sessions
            output.append(f"*[Summary — {len([e for e in exchanges if e[0] == 'user'])} user turns]*\n")
            summarized = summarize_exchanges(exchanges)
            for role, text in summarized:
                if role == "user":
                    output.append(f"- **User**: {text[:200]}")
            output.append("")
        else:
            # Full detail for recent sessions
            turn_num = 0
            for role, text in exchanges:
                if role == "user":
                    turn_num += 1
                    output.append(f"### Turn {turn_num} — User\n")
                    output.append(text)
                    output.append("")
                elif role == "assistant":
                    output.append(f"#### Assistant Response\n")
                    output.append(text)
                    output.append("")
                elif role == "tool_result":
                    output.append(f"> {text[:200]}")
                    output.append("")
                elif role == "system":
                    output.append(f"*[System: {text[:200]}]*")
                    output.append("")

    return "\n".join(output)


def find_project_sessions(project_dir):
    """Find all session JSONL files for a project directory.

    Searches Claude Code, Codex, and OpenCode session directories.
    Returns list of (filepath, mtime) sorted by mtime (oldest first).
    """
    project_dir = os.path.realpath(project_dir)
    sessions = []

    # Claude Code: ~/.claude/projects/<encoded-path>/*.jsonl
    claude_home = Path.home() / ".claude" / "projects"
    if claude_home.exists():
        # Encode path: /Users/tom/myproject → -Users-tom-myproject
        encoded = "-" + project_dir.lstrip("/").replace("/", "-")
        project_path = claude_home / encoded
        if project_path.exists():
            for f in project_path.glob("*.jsonl"):
                # Skip subagent logs (they're in subdirectories)
                if f.parent == project_path:
                    sessions.append((str(f), f.stat().st_mtime))

    # Codex: ~/.codex/sessions/*.jsonl
    codex_path = Path.home() / ".codex" / "sessions"
    if codex_path.exists():
        for f in codex_path.glob("*.jsonl"):
            sessions.append((str(f), f.stat().st_mtime))

    # OpenCode: platform-dependent
    if sys.platform == "darwin":
        oc_path = Path.home() / "Library" / "Application Support" / "opencode" / "sessions"
    else:
        oc_path = Path.home() / ".local" / "share" / "opencode" / "sessions"
    if oc_path.exists():
        for f in oc_path.glob("*.jsonl"):
            sessions.append((str(f), f.stat().st_mtime))

    # Sort by modification time, oldest first
    sessions.sort(key=lambda x: x[1])
    return sessions


def main():
    if len(sys.argv) < 2:
        print(f"Usage:", file=sys.stderr)
        print(f"  {sys.argv[0]} --project-dir /path/to/project", file=sys.stderr)
        print(f"  {sys.argv[0]} /path/to/session.jsonl", file=sys.stderr)
        sys.exit(1)

    if sys.argv[1] == "--project-dir":
        if len(sys.argv) < 3:
            print("Error: --project-dir requires a path argument", file=sys.stderr)
            sys.exit(1)
        project_dir = sys.argv[2]
        if not Path(project_dir).is_dir():
            print(f"Error: {project_dir} is not a directory", file=sys.stderr)
            sys.exit(1)

        sessions = find_project_sessions(project_dir)
        if not sessions:
            print(f"Error: No session files found for project {project_dir}", file=sys.stderr)
            print(f"Searched in ~/.claude/projects/, ~/.codex/sessions/, and opencode sessions", file=sys.stderr)
            sys.exit(1)

        print(f"Found {len(sessions)} session(s)", file=sys.stderr)

        # Parse all sessions
        session_data = []
        for filepath, mtime in sessions:
            exchanges = parse_single_file(filepath)
            if exchanges:  # Skip empty sessions
                session_data.append((filepath, mtime, exchanges))

        if not session_data:
            print("Error: All session files were empty or unparseable", file=sys.stderr)
            sys.exit(1)

        print(f"Parsed {len(session_data)} non-empty session(s)", file=sys.stderr)
        print(build_markdown_multi(session_data))

    else:
        # Single file mode
        jsonl_path = sys.argv[1]
        if not Path(jsonl_path).exists():
            print(f"Error: {jsonl_path} not found", file=sys.stderr)
            sys.exit(1)

        exchanges = parse_single_file(jsonl_path)
        output = ["# Session Transcript\n"]
        turn_num = 0
        for role, text in exchanges:
            if role == "user":
                turn_num += 1
                output.append(f"## Turn {turn_num} — User\n")
                output.append(text)
                output.append("")
            elif role == "assistant":
                output.append(f"### Assistant Response\n")
                output.append(text)
                output.append("")
            elif role == "tool_result":
                output.append(f"> {text[:200]}")
                output.append("")
            elif role == "system":
                output.append(f"*[System: {text[:200]}]*")
                output.append("")
        print("\n".join(output))


if __name__ == "__main__":
    main()

File: skills/retrospect/scripts/parse_session.js
#!/usr/bin/env node
/**
 * Parse AI coding assistant session JSONL files into readable Markdown transcripts.
 *
 * Supports: Claude Code, Codex (OpenAI), OpenCode, and generic JSONL chat formats.
 *
 * Usage:
 *   # Parse all sessions for a project (auto-discovers JSONL files):
 *   node parse_session.js --project-dir /path/to/project
 *
 *   # Parse a single JSONL file:
 *   node parse_session.js /path/to/session.jsonl
 *
 * Outputs Markdown to stdout. Redirect to a file as needed.
 */

const fs = require("fs");
const path = require("path");
const os = require("os");

const FULL_DETAIL_SESSIONS = 5;
const MAX_SUMMARY_TURNS = 10;

function extractTextFromContent(content) {
  if (typeof content === "string") return content.trim();
  if (!Array.isArray(content)) return "";

  const parts = [];
  for (const block of content) {
    if (!block || typeof block !== "object") continue;
    const btype = block.type || "";

    if (btype === "text") {
      parts.push(block.text.trim());
    } else if (btype === "tool_use") {
      const name = block.name || "unknown";
      const inp = block.input || {};
      if (["Read", "Glob", "Grep", "read_file", "grep", "glob"].includes(name)) {
        const target = inp.file_path || inp.pattern || inp.path || "";
        parts.push(`[Tool: ${name}(${target})]`);
      } else if (["Bash", "shell", "run_command"].includes(name)) {
        let cmd = inp.command || "";
        if (cmd.length > 200) cmd = cmd.slice(0, 200) + "...";
        parts.push(`[Tool: ${name}] ${cmd}`);
      } else if (["Write", "Edit", "write_file", "edit_file"].includes(name)) {
        parts.push(`[Tool: ${name}(${inp.file_path || ""})]`);
      } else if (["Agent", "spawn_agent"].includes(name)) {
        const desc = (inp.description || inp.prompt || "").slice(0, 100);
        parts.push(`[Tool: Agent(${desc})]`);
      } else {
        parts.push(`[Tool: ${name}]`);
      }
    } else if (btype === "tool_result") {
      const rc = block.content || "";
      if (typeof rc === "string" && rc.trim()) {
        let text = rc.trim();
        if (text.length > 300) text = text.slice(0, 300) + "... [truncated]";
        parts.push(`[Tool Result] ${text}`);
      } else if (Array.isArray(rc)) {
        for (const sub of rc) {
          if (sub && sub.type === "text") {
            let text = sub.text.trim();
            if (text.length > 300) text = text.slice(0, 300) + "... [truncated]";
            parts.push(`[Tool Result] ${text}`);
          }
        }
      }
    } else if (btype === "function_call") {
      parts.push(`[Function: ${block.name || "unknown"}]`);
    }
    // Skip "thinking" blocks
  }
  return parts.join("\n");
}

function detectFormat(obj) {
  if (obj.type && ["user", "assistant", "system", "file-history-snapshot"].includes(obj.type)) {
    return "claude-code";
  }
  if (obj.role && obj.content !== undefined) return "openai";
  return "unknown";
}

function parseClaudeCode(lines) {
  const exchanges = [];
  for (const line of lines) {
    let obj;
    try { obj = JSON.parse(line); } catch { continue; }

    const msgType = obj.type || "";
    if (["file-history-snapshot", "attachment"].includes(msgType)) continue;
    if (obj.isMeta) continue;

    if (msgType === "user") {
      const content = (obj.message || {}).content || "";
      const text = extractTextFromContent(content);
      if (text && text.includes("<command-name>")) continue;
      if (text && !text.startsWith("[Tool Result]")) {
        exchanges.push(["user", text]);
      } else if (text) {
        exchanges.push(["tool_result", text]);
      }
    } else if (msgType === "assistant") {
      const content = (obj.message || {}).content || [];
      const text = extractTextFromContent(content);
      if (text) exchanges.push(["assistant", text]);
    } else if (msgType === "system") {
      const content = (obj.content || "").trim();
      if (content.length > 10) {
        const subtype = obj.subtype || "";
        if (subtype !== "local_command") {
          exchanges.push(["system", content.slice(0, 500)]);
        }
      }
    }
  }
  return exchanges;
}

function parseOpenAIFormat(lines) {
  const exchanges = [];
  for (const line of lines) {
    let obj;
    try { obj = JSON.parse(line); } catch { continue; }

    const role = obj.role || "";
    const text = extractTextFromContent(obj.content || "");
    if (!text) continue;

    if (["user", "human"].includes(role)) {
      exchanges.push(["user", text]);
    } else if (["assistant", "model"].includes(role)) {
      exchanges.push(["assistant", text]);
    } else if (role === "system" && text.length > 10) {
      exchanges.push(["system", text.slice(0, 500)]);
    } else if (role === "tool" && text.length > 10) {
      exchanges.push(["tool_result", text.slice(0, 300)]);
    }
  }
  return exchanges;
}

function parseSingleFile(filePath) {
  const lines = fs.readFileSync(filePath, "utf-8").trim().split("\n");

  let fmt = "unknown";
  for (const line of lines) {
    try {
      const obj = JSON.parse(line);
      fmt = detectFormat(obj);
      if (fmt !== "unknown") break;
    } catch { continue; }
  }

  if (fmt === "claude-code") return parseClaudeCode(lines);
  if (fmt === "openai") return parseOpenAIFormat(lines);

  const cc = parseClaudeCode(lines);
  const oai = parseOpenAIFormat(lines);
  return cc.length >= oai.length ? cc : oai;
}

function summarizeExchanges(exchanges) {
  const summary = [];
  let count = 0;
  for (const [role, text] of exchanges) {
    if (role === "user") {
      count++;
      if (count > MAX_SUMMARY_TURNS) {
        summary.push(["user", `... [${count}+ more user turns omitted] ...`]);
        break;
      }
      summary.push(["user", text.length > 300 ? text.slice(0, 300) + "... [truncated]" : text]);
    }
  }
  return summary;
}

function buildMarkdownMulti(sessionList) {
  const output = ["# Project Session Transcripts\n"];
  const total = sessionList.length;
  const fullStart = Math.max(0, total - FULL_DETAIL_SESSIONS);

  for (let i = 0; i < sessionList.length; i++) {
    const { file, mtime, exchanges } = sessionList[i];
    if (!exchanges.length) continue;

    const sessionNum = i + 1;
    const dateStr = new Date(mtime).toISOString().replace("T", " ").slice(0, 16);
    const sessionId = path.basename(file, ".jsonl").slice(0, 8);

    output.push("---\n");
    output.push(`## Session ${sessionNum} (${dateStr}, id: ${sessionId}...)\n`);

    if (i < fullStart) {
      const userCount = exchanges.filter(([r]) => r === "user").length;
      output.push(`*[Summary — ${userCount} user turns]*\n`);
      for (const [role, text] of summarizeExchanges(exchanges)) {
        if (role === "user") output.push(`- **User**: ${text.slice(0, 200)}`);
      }
      output.push("");
    } else {
      let turnNum = 0;
      for (const [role, text] of exchanges) {
        if (role === "user") {
          turnNum++;
          output.push(`### Turn ${turnNum} — User\n`);
          output.push(text, "");
        } else if (role === "assistant") {
          output.push("#### Assistant Response\n");
          output.push(text, "");
        } else if (role === "tool_result") {
          output.push(`> ${text.slice(0, 200)}`, "");
        } else if (role === "system") {
          output.push(`*[System: ${text.slice(0, 200)}]*`, "");
        }
      }
    }
  }
  return output.join("\n");
}

function findProjectSessions(projectDir) {
  projectDir = fs.realpathSync(projectDir);
  const sessions = [];

  // Claude Code: ~/.claude/projects/<encoded-path>/*.jsonl
  const claudeHome = path.join(os.homedir(), ".claude", "projects");
  if (fs.existsSync(claudeHome)) {
    const encoded = "-" + projectDir.replace(/^\//,"").replace(/\//g, "-");
    const projectPath = path.join(claudeHome, encoded);
    if (fs.existsSync(projectPath)) {
      for (const f of fs.readdirSync(projectPath)) {
        if (!f.endsWith(".jsonl")) continue;
        const full = path.join(projectPath, f);
        // Skip subagent logs (only top-level files)
        if (fs.statSync(full).isFile()) {
          sessions.push({ file: full, mtime: fs.statSync(full).mtimeMs });
        }
      }
    }
  }

  // Codex: ~/.codex/sessions/*.jsonl
  const codexPath = path.join(os.homedir(), ".codex", "sessions");
  if (fs.existsSync(codexPath)) {
    for (const f of fs.readdirSync(codexPath)) {
      if (!f.endsWith(".jsonl")) continue;
      const full = path.join(codexPath, f);
      if (fs.statSync(full).isFile()) {
        sessions.push({ file: full, mtime: fs.statSync(full).mtimeMs });
      }
    }
  }

  // OpenCode: platform-dependent
  const ocPath = process.platform === "darwin"
    ? path.join(os.homedir(), "Library", "Application Support", "opencode", "sessions")
    : path.join(os.homedir(), ".local", "share", "opencode", "sessions");
  if (fs.existsSync(ocPath)) {
    for (const f of fs.readdirSync(ocPath)) {
      if (!f.endsWith(".jsonl")) continue;
      const full = path.join(ocPath, f);
      if (fs.statSync(full).isFile()) {
        sessions.push({ file: full, mtime: fs.statSync(full).mtimeMs });
      }
    }
  }

  sessions.sort((a, b) => a.mtime - b.mtime);
  return sessions;
}

function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.error("Usage:");
    console.error("  node parse_session.js --project-dir /path/to/project");
    console.error("  node parse_session.js /path/to/session.jsonl");
    process.exit(1);
  }

  if (args[0] === "--project-dir") {
    const projectDir = args[1];
    if (!projectDir || !fs.existsSync(projectDir) || !fs.statSync(projectDir).isDirectory()) {
      console.error(`Error: ${projectDir || "(missing)"} is not a directory`);
      process.exit(1);
    }

    const sessions = findProjectSessions(projectDir);
    if (!sessions.length) {
      console.error(`Error: No session files found for project ${projectDir}`);
      console.error("Searched in ~/.claude/projects/, ~/.codex/sessions/, and opencode sessions");
      process.exit(1);
    }

    console.error(`Found ${sessions.length} session(s)`);

    const sessionData = [];
    for (const { file, mtime } of sessions) {
      const exchanges = parseSingleFile(file);
      if (exchanges.length) sessionData.push({ file, mtime, exchanges });
    }

    if (!sessionData.length) {
      console.error("Error: All session files were empty or unparseable");
      process.exit(1);
    }

    console.error(`Parsed ${sessionData.length} non-empty session(s)`);
    console.log(buildMarkdownMulti(sessionData));

  } else {
    const filePath = args[0];
    if (!fs.existsSync(filePath)) {
      console.error(`Error: ${filePath} not found`);
      process.exit(1);
    }

    const exchanges = parseSingleFile(filePath);
    const output = ["# Session Transcript\n"];
    let turnNum = 0;
    for (const [role, text] of exchanges) {
      if (role === "user") {
        turnNum++;
        output.push(`## Turn ${turnNum} — User\n`, text, "");
      } else if (role === "assistant") {
        output.push("### Assistant Response\n", text, "");
      } else if (role === "tool_result") {
        output.push(`> ${text.slice(0, 200)}`, "");
      } else if (role === "system") {
        output.push(`*[System: ${text.slice(0, 200)}]*`, "");
      }
    }
    console.log(output.join("\n"));
  }
}

main();
